stdlib/src/tempfile/tempfile.mojo

```diff
@@ -239,3 +239,137 @@ struct TemporaryDirectory:
             return True
         except:
             return False
+
+
+struct NamedTemporaryFile:
+    """A handle to a temporary file."""
+
+    var _file_handle: FileHandle
+    """The underlying file handle."""
+    var _delete: Bool
+    """Whether the file is deleted on close."""
+    var name: String
+    """Name of the file."""
+
+    fn __init__(
+        inout self,
+        mode: String = "w",
+        suffix: String = "",
+        prefix: String = "tmp",
+        dir: Optional[String] = None,
+        delete: Bool = True,
+    ) raises:
+        """Create a named temporary file. Can be used as a context manager.
+        This is a wrapper around a `FileHandle`,
+        os.remove is called in close method if `delete` is True.
+
+        Args:
+            mode: The mode to open the file in (the mode can be "r" or "w").
+            suffix: Suffix to use for the file name.
+            prefix: Prefix to use for the file name.
+            dir: Directory in which the file will be created.
+            delete: Whether the file is deleted on close.
+        """
+
+        var final_dir = Path(dir.value()) if dir else Path(
+            _get_default_tempdir()
+        )
+
+        self._delete = delete
+        self.name = ""
+
+        for _ in range(TMP_MAX):
+            var potential_name = final_dir / (
+                prefix + _get_random_name() + suffix
+            )
+            if os.path.exists(potential_name):
+                continue
+            try:
+                # TODO for now this name could be relative,
+                # python implementation expands the path,
+                # but several functions are not yet implemented in mojo
+                # i.e. abspath, normpath
```

Comment:
Followup idea: Maybe it would be better to do this earlier than later, given that it would be already useful in a few places.

Reply:
yes, I don't think I will have some time soon, but I opened two new issues to implement those functions: #2961 and #2962